File: API/2153_Anchored_Momentum_Candle/CS/AnchoredMomentumCandleStrategy.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Ecng.Common;
using Ecng.Collections;
using Ecng.Serialization;
using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Strategy based on the Anchored Momentum Candle indicator.
/// Compares open-based and close-based momentum to determine candle color.
/// </summary>
public class AnchoredMomentumCandleStrategy : Strategy
{
	private readonly StrategyParam<int> _momPeriod;
	private readonly StrategyParam<int> _smoothPeriod;
	private readonly StrategyParam<DataType> _candleType;

	private readonly Queue<decimal> _openQueue = new();
	private readonly Queue<decimal> _closeQueue = new();
	private decimal _sumOpen, _sumClose;
	private decimal _emaOpen, _emaClose;
	private bool _emaInit;
	private decimal? _prevColor;

	public int MomPeriod { get => _momPeriod.Value; set => _momPeriod.Value = value; }
	public int SmoothPeriod { get => _smoothPeriod.Value; set => _smoothPeriod.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public AnchoredMomentumCandleStrategy()
	{
		_momPeriod = Param(nameof(MomPeriod), 8)
			.SetGreaterThanZero()
			.SetDisplay("Momentum Period", "SMA length", "Parameters")
			.SetOptimize(2, 20, 1);

		_smoothPeriod = Param(nameof(SmoothPeriod), 6)
			.SetGreaterThanZero()
			.SetDisplay("Smooth Period", "EMA length", "Parameters")
			.SetOptimize(2, 20, 1);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Working timeframe", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_openQueue.Clear();
		_closeQueue.Clear();
		_sumOpen = _sumClose = 0m;
		_emaOpen = _emaClose = 0m;
		_emaInit = false;
		_prevColor = null;
	}

	/// <inheritdoc />
	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(ProcessCandle).Start();
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var open = candle.OpenPrice;
		var close = candle.ClosePrice;

		_sumOpen += open;
		_openQueue.Enqueue(open);
		if (_openQueue.Count > MomPeriod)
			_sumOpen -= _openQueue.Dequeue();

		_sumClose += close;
		_closeQueue.Enqueue(close);
		if (_closeQueue.Count > MomPeriod)
			_sumClose -= _closeQueue.Dequeue();

		var k = 2m / (SmoothPeriod + 1);

		if (!_emaInit)
		{
			_emaOpen = open;
			_emaClose = close;
			_emaInit = true;
		}
		else
		{
			_emaOpen = k * open + (1 - k) * _emaOpen;
			_emaClose = k * close + (1 - k) * _emaClose;
		}

		if (_openQueue.Count < MomPeriod)
			return;

		var smaOpen = _sumOpen / MomPeriod;
		var smaClose = _sumClose / MomPeriod;

		var openMomentum = smaOpen == 0m ? 0m : 100m * (_emaOpen / smaOpen - 1m);
		var closeMomentum = smaClose == 0m ? 0m : 100m * (_emaClose / smaClose - 1m);

		var color = openMomentum < closeMomentum ? 2m : openMomentum > closeMomentum ? 0m : 1m;

		if (_prevColor == null)
		{
			_prevColor = color;
			return;
		}

		if (color == 2m && _prevColor != 2m)
		{
			if (Position < 0)
				BuyMarket();
			if (Position <= 0)
				BuyMarket();
		}
		else if (color == 0m && _prevColor != 0m)
		{
			if (Position > 0)
				SellMarket();
			if (Position >= 0)
				SellMarket();
		}

		_prevColor = color;
	}
}

File: API/2101-2200/2153_Anchored_Momentum_Candle/PY/anchored_momentum_candle_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Strategies import Strategy


class anchored_momentum_candle_strategy(Strategy):
    def __init__(self):
        super(anchored_momentum_candle_strategy, self).__init__()
        self._mom_period = self.Param("MomPeriod", 8) \
            .SetDisplay("Momentum Period", "SMA length", "Parameters")
        self._smooth_period = self.Param("SmoothPeriod", 6) \
            .SetDisplay("Smooth Period", "EMA length", "Parameters")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Working timeframe", "General")
        self._open_queue = []
        self._close_queue = []
        self._sum_open = 0.0
        self._sum_close = 0.0
        self._ema_open = 0.0
        self._ema_close = 0.0
        self._ema_init = False
        self._prev_color = None

    @property
    def mom_period(self):
        return self._mom_period.Value

    @property
    def smooth_period(self):
        return self._smooth_period.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(anchored_momentum_candle_strategy, self).OnReseted()
        self._open_queue = []
        self._close_queue = []
        self._sum_open = 0.0
        self._sum_close = 0.0
        self._ema_open = 0.0
        self._ema_close = 0.0
        self._ema_init = False
        self._prev_color = None

    def OnStarted2(self, time):
        super(anchored_momentum_candle_strategy, self).OnStarted2(time)
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(self.process_candle).Start()

    def process_candle(self, candle):
        if candle.State != CandleStates.Finished:
            return
        open_p = float(candle.OpenPrice)
        close_p = float(candle.ClosePrice)
        mp = int(self.mom_period)
        sp = int(self.smooth_period)
        self._sum_open += open_p
        self._open_queue.append(open_p)
        if len(self._open_queue) > mp:
            self._sum_open -= self._open_queue.pop(0)
        self._sum_close += close_p
        self._close_queue.append(close_p)
        if len(self._close_queue) > mp:
            self._sum_close -= self._close_queue.pop(0)
        k = 2.0 / (sp + 1)
        if not self._ema_init:
            self._ema_open = open_p
            self._ema_close = close_p
            self._ema_init = True
        else:
            self._ema_open = k * open_p + (1.0 - k) * self._ema_open
            self._ema_close = k * close_p + (1.0 - k) * self._ema_close
        if len(self._open_queue) < mp:
            return
        sma_open = self._sum_open / mp
        sma_close = self._sum_close / mp
        open_mom = 100.0 * (self._ema_open / sma_open - 1.0) if sma_open != 0 else 0.0
        close_mom = 100.0 * (self._ema_close / sma_close - 1.0) if sma_close != 0 else 0.0
        if open_mom < close_mom:
            color = 2.0
        elif open_mom > close_mom:
            color = 0.0
        else:
            color = 1.0
        if self._prev_color is None:
            self._prev_color = color
            return
        if color == 2.0 and self._prev_color != 2.0:
            if self.Position < 0:
                self.BuyMarket()
            if self.Position <= 0:
                self.BuyMarket()
        elif color == 0.0 and self._prev_color != 0.0:
            if self.Position > 0:
                self.SellMarket()
            if self.Position >= 0:
                self.SellMarket()
        self._prev_color = color

    def CreateClone(self):
        return anchored_momentum_candle_strategy()
